packages/insured-bridge-relayer/src/ProfitabilityCalculator.ts

```diff
@@ -0,0 +1,227 @@
+      }
+    }
+
+    // Fetch the prices of each token, denominated in ETH.
+    const tokenPrices = await Promise.all(
+      this.l1Tokens.map((l1Token) => this.coingecko.getCurrentPriceByContract(l1Token, "eth"))
+    );
+
+    // For each token, extract the price and convert to a wei'd BN. Note that if the type is WETH then the price is 1.
+    // This is done as coingecko does not always return the price of 1 for WETH, in ETH.
+    for (const [index, priceResponse] of tokenPrices.entries()) {
+      this.l1TokenInfo[this.l1Tokens[index]].tokenEthPrice =
+        this.l1TokenInfo[this.l1Tokens[index]].tokenType == TokenType.WETH ? toBNWei("1") : toBNWei(priceResponse[1]);
+    }
+    this.logger.debug({
+      at: "ProfitabilityCalculator",
+      message: "Updated prices",
+      tokenInfo: objectMap(this.l1TokenInfo, (value: { tokenType: TokenType; tokenEthPrice: BN }) => {
+        return { tokenType: TokenType[value.tokenType], tokenEthPrice: fromWei(value.tokenEthPrice) };
+      }),
+    });
+  }
+
+  getRelaySubmitTypeBasedOnProfitability(
+    l1Token: string,
+    cumulativeGasPrice: BN, // sum of the most recent base fee per gas and the current estimated maxPriorityFeePerGas.
+    slowRevenue: BN,
+    speedUpRevenue: BN,
+    instantRevenue: BN
+  ): RelaySubmitType {
+    this._throwIfNotInitialized();
+    if (!this.l1TokenInfo[l1Token]) throw new Error("Token info not found. Ensure to construct correctly");
+    const { tokenType, tokenEthPrice } = this.l1TokenInfo[l1Token];
+
+    this.logger.debug({
+      at: "ProfitabilityCalculator",
+      message: "Considering relay profitability",
+      l1Token,
+      tokenType: TokenType[tokenType],
+      tokenEthPrice: fromWei(tokenEthPrice),
+      cumulativeGasPrice: cumulativeGasPrice.toString(),
+      relayerDiscount: fromWei(this.relayerDiscount),
+      slowRevenue: slowRevenue.toString(),
+      speedUpRevenue: speedUpRevenue.toString(),
+      instantRevenue: instantRevenue.toString(),
+    });
+
+    // Calculate the transaction submission costs depending on the gas price and the token type being relayed. This
+    // considers the relayer discount, applied to all relay actions equally.
+    const ethSubmissionCost = this.getRelayEthSubmissionCost(cumulativeGasPrice, tokenType);
+
+    // Calculate the expected revenue, in ETH, based on the amount of tokens being offered in fees.
+    const ethRevenue = this.getEthRevenue(tokenEthPrice, slowRevenue, speedUpRevenue, instantRevenue);
+
+    // Calculate the relay profitability as the difference between revenue and cost.
+    const ethProfitability = this.getProfit(ethSubmissionCost, ethRevenue);
+
+    // Finally, determine the relay submit type based on the profitability.
+    let relaySubmitType: RelaySubmitType;
+    if (
+      ethProfitability.instantEthProfit.gt(toBN(0)) &&
+      ethProfitability.instantEthProfit.gt(ethProfitability.speedUpEthProfit) &&
```

Comment:
nit: should these be `gte` or `gt` you think? I don't really know myself, but currently it seems like we prefer slow relays all things being equal, but shouldn't we prefer instant relays? I guess users would prefer our bots to prefer instant relays, while relayers would rather be slow?

Reply:
I think `gte` just on the instant makes sense as you are right we should always instant relay if posible.

Reply:
agreed